### PR TITLE
Facet float-integer becomes facet number

### LINKED ISSUE #142: Only two facet types: string and number
_Related to this specification: https://github.com/meilisearch/specifications/pull/24_

Milli needs to know the facet types.
Currently, Milli handles 3 types: float, integer and string.

Milli will handle 2 types only: number and string.
Numbers are any integers or floats, and these will be converted into float.

### PR DESCRIPTION
**Pull request by Kerollmops**
This pull request fixes #142 by converting all the integers into floats, introducing a new "number" type. The exact precision for integers converted into [64bits floats is between -2^53 and 2^53](https://en.wikipedia.org/wiki/Double-precision_floating-point_format#Precision_limitations_on_integer_values), it has been decided that it was enough for common use cases and ease the search engine usage.

More information can be found [in this specification draft document](https://hackmd.io/@WZKoavtcShyXvCB_qthGRQ/Hka6af-Vu).

- [x] Convert booleans into the appropriate facet type when declared. Keep in mind that the original document is never modified, the document is always returned as it was received.
- [x] Don't skip empty strings and consider them as valid facet values.
- [x] Remove trait bounds from the facet functions and types (`FacetIter`, `FacetRange`) to work only for numbers (`f64`).
- [x] Warn instead of raising an error and stop the indexing when a conversion to the facet type is impossible.

---

In a future pull request, we should split both facet databases of strings and floats. Doing so would make it possible to use a different implementation of facet levels for strings and numbers, strings can't work like numbers due to their inherent unknown size making facet distribution aggregation for strings slow. Then wrap both facet databases into a type that would help us work with facets strings and numbers. There can be helping methods to return iterators on ranges of numbers, iterators for facet ranges...

### CODE REVIEW

**Comment by curquiza** [comment]
Will this PR fix completely https://github.com/meilisearch/milli/issues/142 or this issue will need another PR?

**Comment by Kerollmops** [comment]
@curquiza Yep, this fixes #144.

**Comment by Kerollmops** [comment]
Hey @MarinPostma, Could you please check that the distinct system is still valid? Thank you :)

File: milli/src/heed_codec/facet/mod.rs [hunk 1]
```diff
@@ -1,13 +1,9 @@
 mod facet_level_value_f64_codec;
-mod facet_level_value_i64_codec;
 mod facet_value_string_codec;
 mod field_doc_id_facet_f64_codec;
-mod field_doc_id_facet_i64_codec;
 mod field_doc_id_facet_string_codec;
 
 pub use self::facet_level_value_f64_codec::FacetLevelValueF64Codec;
```
Comment (MarinPostma) [inline milli/src/heed_codec/facet/mod.rs:6]: Maybe we should change this to `FacetLevelValueNumberCodec`

**Review by MarinPostma — CHANGES_REQUESTED** [review]
Looks good to me on the distinct side.

So no inference for now, we just drop the `Integer` and `Float` facet types and replace them with `Number`?

File: milli/src/heed_codec/facet/mod.rs (hunk 1 above)
Comment (Kerollmops) [reply milli/src/heed_codec/facet/mod.rs:6]: As this is a low-level codec implementation that encodes and decodes f64 directly, I prefer keeping the explicit name.

**Comment by Kerollmops** [comment]
> So no inference for now, we just drop the `Integer` and `Float` facet types and replace them with `Number`?

The inference is no more needed, it has been decided that there are only soft type conversions, strings will be moved into the string database and JSON numbers will be moved in the f64 database.

**Pull request merged**